helix-term/src/commands.rs

```diff
@@ -3975,40 +3987,55 @@ pub fn completion(cx: &mut Context) {
     let offset = iter.take_while(|ch| chars::char_is_word(*ch)).count();
     let start_offset = cursor.saturating_sub(offset);
 
-    cx.callback(
-        future,
-        move |editor, compositor, response: Option<lsp::CompletionResponse>| {
-            if editor.mode != Mode::Insert {
-                // we're not in insert mode anymore
-                return;
-            }
+    let mut requests = Vec::new();
 
-            let items = match response {
-                Some(lsp::CompletionResponse::Array(items)) => items,
-                // TODO: do something with is_incomplete
-                Some(lsp::CompletionResponse::List(lsp::CompletionList {
-                    is_incomplete: _is_incomplete,
-                    items,
-                })) => items,
-                None => Vec::new(),
-            };
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::Completion) {
+        let language_server_id = language_server.id();
 
-            if items.is_empty() {
-                // editor.set_error("No completion available");
-                return;
-            }
-            let size = compositor.size();
-            let ui = compositor.find::<ui::EditorView>().unwrap();
-            ui.set_completion(
-                editor,
-                items,
-                offset_encoding,
-                start_offset,
-                trigger_offset,
-                size,
-            );
-        },
-    );
+        let offset_encoding = language_server.offset_encoding();
+
+        let pos = pos_to_lsp_pos(doc.text(), cursor, offset_encoding);
+
+        let future = match language_server.completion(doc.identifier(), pos, None) {
+            Some(future) => future,
+            None => return,
```

Comment:
this return can drop previous requests. language servers might be configured with completion flag but actually not have the capabilities. as `doc.language_servers_with_feature()` only filters by flags from *languages.toml* but `language_server.completion()` checks the actual capabilities and might not return a future, effectively canceling previous requests from this loop.

consider this language configuration:
```toml
language-servers = [ "vscode-scss", "vscode-css" ]
```
*vscode-scss* has a CompletionProvider but *vscode-css* doesn't, `completion()` returns no future and the completion request from *vscode-scss* is canceled. no completion menu appears.

maybe like this, so all requests actually go off
```rust
if let Some(future) = language_server.completion(doc.identifier(), pos, None) {
    requests.push((future, language_server_id, offset_encoding));
}
```

Reply:
Whoops, good catch that is indeed not intended, I fixed this and improved similar code areas (while they were not incorrect, as handled with a `continue` instead of a `return`, your solution seems more idiomatic.